##Data Preparation

Unduh data csv dari: https://github.com/realpython/python-data-cleaning/blob/master/Datasets/BL-Flickr-Images-Book.csv

1. Import Library yang dibutuhkan

In [4]:
import pandas as pd
import numpy as np

df = pd.read_csv("BL-Flickr-Images-Book.csv")

In [5]:
print(df.dtypes)

Identifier                  int64
Edition Statement          object
Place of Publication       object
Date of Publication        object
Publisher                  object
Title                      object
Author                     object
Contributors               object
Corporate Author          float64
Corporate Contributors    float64
Former owner               object
Engraver                  float64
Issuance type              object
Flickr URL                 object
Shelfmarks                 object
dtype: object


In [6]:
df.describe()

Identifier  Corporate Author  Corporate Contributors  Engraver
count  8.287000e+03               0.0                     0.0       0.0
mean   2.017344e+06               NaN                     NaN       NaN
std    1.190379e+06               NaN                     NaN       NaN
min    2.060000e+02               NaN                     NaN       NaN
25%    9.157875e+05               NaN                     NaN       NaN
50%    2.043707e+06               NaN                     NaN       NaN
75%    3.047430e+06               NaN                     NaN       NaN
max    4.160339e+06               NaN                     NaN       NaN

In [7]:
df.head()

Identifier  ...                           Shelfmarks
0         206  ...    British Library HMNTS 12641.b.30.
1         216  ...    British Library HMNTS 12626.cc.2.
2         218  ...    British Library HMNTS 12625.dd.1.
3         472  ...  British Library HMNTS 10369.bbb.15.
4         480  ...     British Library HMNTS 9007.d.28.

[5 rows x 15 columns]

In [8]:
yg_dihapus = ['Edition Statement', 'Corporate Author']
df.drop(yg_dihapus, inplace =True, axis =1)

In [9]:
df.head()

Identifier  ...                           Shelfmarks
0         206  ...    British Library HMNTS 12641.b.30.
1         216  ...    British Library HMNTS 12626.cc.2.
2         218  ...    British Library HMNTS 12625.dd.1.
3         472  ...  British Library HMNTS 10369.bbb.15.
4         480  ...     British Library HMNTS 9007.d.28.

[5 rows x 13 columns]

2. Mengubah Indeks di data

In [10]:
df['Identifier'].is_unique

True

In [11]:
df = df.set_index('Identifier')

In [12]:
df.head()

Place of Publication  ...                           Shelfmarks
Identifier                            ...                                     
206                           London  ...    British Library HMNTS 12641.b.30.
216         London; Virtue & Yorston  ...    British Library HMNTS 12626.cc.2.
218                           London  ...    British Library HMNTS 12625.dd.1.
472                           London  ...  British Library HMNTS 10369.bbb.15.
480                           London  ...     British Library HMNTS 9007.d.28.

[5 rows x 12 columns]

In [13]:
df.loc[206]

Place of Publication                                                 London
Date of Publication                                             1879 [1878]
Publisher                                                  S. Tinsley & Co.
Title                                     Walter Forbes. [A novel.] By A. A
Author                                                                A. A.
Contributors                                                FORBES, Walter.
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                                British Library HMNTS 12641.b.30.
Name: 206, dtype: object

In [14]:
df.loc[472]

Place of Publication                                                 London
Date of Publication                                                    1851
Publisher                                                     James Darling
Title                     Welsh Sketches, chiefly ecclesiastical, to the...
Author                                                            A., E. S.
Contributors                                    Appleyard, Ernest Silvanus.
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                              British Library HMNTS 10369.bbb.15.
Name: 472, dtype: object

3. Merapikan Fields

In [15]:
df.dtypes.value_counts()

object     10
float64     2
dtype: int64

In [16]:
df.loc[1905:, 'Date of Publication'].head(15)

Identifier
1905                  1888
1929           1839, 38-54
2836                  1897
2854                  1865
2956               1860-63
2957                  1873
3017                  1866
3131                  1899
4598                  1814
4884                  1820
4976                  1800
5382    1847, 48 [1846-48]
5385               [1897?]
5389               [1897?]
5432                  1893
Name: Date of Publication, dtype: object

~hilangkan tanggal lain dalam kurung siku
~konversi rentang tanggal
~hilangkan tanggal yang gak jelas [1897?]--> NaN
~Konversi NaN

In [17]:
#regex = ekspresi reguler, \d = memberi tahu mewakili random digit, {4} = digit yang diamhil
regex = r'^(\d{4})'

In [18]:
ekstrak = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
ekstrak.head(15)

Identifier
206     1879
216     1868
218     1869
472     1851
480     1857
481     1875
519     1872
667      NaN
874     1676
1143    1679
1280    1802
1808    1859
1905    1888
1929    1839
2836    1897
Name: Date of Publication, dtype: object

In [19]:
df['Date of Publication'] = pd.to_numeric(ekstrak)
df['Date of Publication'].dtype

dtype('float64')

4. Membersihkan fields (kolom)

In [20]:
df['Place of Publication'].head(15)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
1280                               Coventry
1808                            Christiania
1905                                Firenze
1929                              Amsterdam
2836                                 Savona
Name: Place of Publication, dtype: object

In [21]:
df.loc[667]

Place of Publication                    pp. 40. G. Bryan & Co: Oxford, 1898
Date of Publication                                                     NaN
Publisher                                                               NaN
Title                     The Coming of Spring, and other poems. By J. A...
Author                                                        A., J.|A., J.
Contributors                                  ANDREWS, J. - Writer of Verse
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                               British Library HMNTS 011652.g.73.
Name: 667, dtype: object

In [22]:
df.loc[4157862]

Place of Publication                                    Newcastle-upon-Tyne
Date of Publication                                                  1867.0
Publisher                                                        T. Fordyce
Title                     Local Records; or, Historical Register of rema...
Author                        FORDYCE, T. - Printer, of Newcastle-upon-Tyne
Contributors               SYKES, John - Bookseller, of Newcastle-upon-Tyne
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                British Library HMNTS|British Library HMNTS 01...
Name: 4157862, dtype: object

In [23]:
df['Place of Publication'].tail(15)

Identifier
4115138    G. Eld: London, 1608
4116063                  London
4117526                 Leipzig
4117583                  London
4117749                New York
4117751                New York
4117752                  London
4156359                New York
4157746                  London
4157862     Newcastle-upon-Tyne
4158088                  London
4158128                   Derby
4159563                  London
4159587     Newcastle upon Tyne
4160339                  London
Name: Place of Publication, dtype: object

In [24]:
df.loc[4115138]

Place of Publication                                   G. Eld: London, 1608
Date of Publication                                                     NaN
Publisher                                                               NaN
Title                     The Revenger's Tagaedie. As it hath beene sund...
Author                                                                  NaN
Contributors                                               TOURNEUR, Cyril.
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                British Library HMNTS 644.c.80.|British Librar...
Name: 4115138, dtype: object

In [25]:
publikasi = df['Place of Publication']
london = publikasi.str.contains('London')
london[:5]

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [26]:
oxford = publikasi.str.contains('Oxford')

In [27]:
df['Place of Publication'] = np.where(london, 'London',
                                      np.where(oxford, 'Oxford', 
                                        publikasi.str.replace('-', ' ')))

In [28]:
df['Place of Publication'].head(15)

Identifier
206          London
216          London
218          London
472          London
480          London
481          London
519          London
667          Oxford
874          London
1143         London
1280       Coventry
1808    Christiania
1905        Firenze
1929      Amsterdam
2836         Savona
Name: Place of Publication, dtype: object

kita gunakan dataset yang lain, unduh di https://github.com/realpython/python-data-cleaning/blob/master/Datasets/university_towns.txt

5. Membersihkan dataset dengan applymap

In [31]:
university_town = []
with open ("university_towns.txt") as file :
    for line in file :
      if '[edit]' in line:
          state = line
      else: 
          university_town.append((state, line))
university_town[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

In [32]:
df_kota = pd.DataFrame(university_town, columns = ['State', 'RegionName'])
df_kota.head(15)

State                                         RegionName
0    Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1    Alabama[edit]\n           Florence (University of North Alabama)\n
2    Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3    Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4    Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n
5    Alabama[edit]\n                        Troy (Troy University)[2]\n
6    Alabama[edit]\n  Tuscaloosa (University of Alabama, Stillman Co...
7    Alabama[edit]\n                Tuskegee (Tuskegee University)[5]\n
8     Alaska[edit]\n    Fairbanks (University of Alaska Fairbanks)[2]\n
9    Arizona[edit]\n       Flagstaff (Northern Arizona University)[6]\n
10   Arizona[edit]\n                 Tempe (Arizona State University)\n
11   Arizona[edit]\n                   Tucson (University of Arizona)\n
12  Arkansas[edit]\n  Arkadelphia (Henderson State University, Ouach...
13  Arkansas[edit]\n  Conway (Central Baptist College, Hendrix Colle...
14  Arkansas[edit]\n         Fayetteville (University of Arkansas)[7]\n

In [33]:
def getCity(item):
    if '(' in item:
        return item[:item.find('(')]
    elif '[' in item:
        return item[:item.find('[')]
    else:
        return item

In [35]:
df_kota = df_kota.applymap(getCity)

In [36]:
df_kota.head()

State     RegionName
0  Alabama        Auburn 
1  Alabama      Florence 
2  Alabama  Jacksonville 
3  Alabama    Livingston 
4  Alabama    Montevallo

In [37]:
df_kota.head(15)

State     RegionName
0    Alabama        Auburn 
1    Alabama      Florence 
2    Alabama  Jacksonville 
3    Alabama    Livingston 
4    Alabama    Montevallo 
5    Alabama          Troy 
6    Alabama    Tuscaloosa 
7    Alabama      Tuskegee 
8     Alaska     Fairbanks 
9    Arizona     Flagstaff 
10   Arizona         Tempe 
11   Arizona        Tucson 
12  Arkansas   Arkadelphia 
13  Arkansas        Conway 
14  Arkansas  Fayetteville